jitsi-media-transform/src/main/kotlin/org/jitsi/nlj/util/IndexTracker.kt

```diff
@@ -0,0 +1,137 @@
+/*
+ * Copyright @ 2018 - present 8x8, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.jitsi.nlj.util
+
+/**
+ * Index tracker inspired by the RFC3711 RTP sequence number index tracker.
+ */
+
+class PictureIdIndexTracker : IntIndexTracker(15)
+class RtpSequenceIndexTracker : IntIndexTracker(16)
+class RtpTimestampIndexTracker : LongIndexTracker(32)
+
+open class IntIndexTracker(val bits: Int) : IndexTracker<Int>() {
+    init {
+        require(bits in 1..31) { "bits must be between 1 and 31" }
+    }
+    override fun addRollover(seqNum: Int, roc: Long) = (1 shl bits) * roc + seqNum
+    override fun rollsOver(a: Int, b: Int): Boolean = isOlderThan(a, b) && b < a
+    override fun isOlderThan(a: Int, b: Int): Boolean = delta(a, b) < 0
+    override fun toLong(t: Int): Long = t.toLong()
+
+    private fun delta(a: Int, b: Int): Int {
+        val diff = a - b
+        return when {
+            diff < -(1 shl (bits - 1)) -> diff + (1 shl bits)
+            diff > (1 shl (bits - 1)) -> diff - (1 shl bits)
+            else -> diff
+        }
+    }
+}
+
+open class LongIndexTracker(val bits: Int) : IndexTracker<Long>() {
+    init {
+        require(bits in 1..63) { "bits must be between 1 and 63" }
+    }
+    override fun addRollover(seqNum: Long, roc: Long) = (1L shl bits) * roc + seqNum
+    override fun rollsOver(a: Long, b: Long): Boolean = isOlderThan(a, b) && b < a
+    override fun isOlderThan(a: Long, b: Long): Boolean = delta(a, b) < 0
+    override fun toLong(t: Long): Long = t
+
+    private fun delta(a: Long, b: Long): Long {
+        val diff = a - b
+        return when {
+            diff < -(1L shl (bits - 1)) -> diff + (1L shl bits)
+            diff > (1L shl (bits - 1)) -> diff - (1L shl bits)
+            else -> diff
+        }
+    }
+}
+
+sealed class IndexTracker<T> {
```

Comment:
Why did you make this sealed?

Reply:
It has to be public because it defines the API, but I want to keep the helper functions internal. I can technically make it just `abstract`, but then you can't define a superclass outside the package because you can't access the internal functions need to be overriden. So it would be effectively sealed except you get a somewhat cryptic error when you try to define a super class: `MyIndexTracker inherits invisible abstract members: internal abstract fun addRollover(seqNum: Int, roc: Long): Long defined in org.jitsi.nlj.util.IndexTracker`